examples/LaravelSampleApp/app/Http/Controllers/GoogleAdsApiController.php

```diff
@@ -76,52 +78,112 @@ public function showReportAction(
         } else {
-            // Loads from the session all the information that is necessary to process the GET
-            // request (report result page).
+            // Loads from the session all the information that is necessary to process any
+            // requests (report result page).
+            $customerId = $request->session()->get('customerId');
             $selectedFields = $request->session()->get('selectedFields');
             $entriesPerPage = $request->session()->get('entriesPerPage');
-            $collection = $request->session()->get('collection');
+            $query = $request->session()->get('query');
+            $pageTokens = $request->session()->get('pageTokens');
         }
 
+        // Determines the number of the page to load (the first one by default).
         $pageNo = $request->input('page') ?: 1;
 
-        // Creates a length aware paginator to supply a given page of results for the view, based on
-        // the collection of results, the page number and the specified number of entries per page.
+        // Fetches pages in sequence to find the page token of the requested page if unknown.
+        while (count($pageTokens) < $pageNo) {
+            // Fetches the page for the latest known page token.
+            $response = $googleAdsClient->getGoogleAdsServiceClient()->search(
+                $customerId,
+                $query,
+                [
+                    'pageSize' => $entriesPerPage,
+                    // Requests to return the total results count. This is necessary to
+                    // determine how many pages of results exist.
+                    'returnTotalResultsCount' => true,
+                    'pageToken' => end($pageTokens)
+                ]
+            );
+            if ($response->getPage()->getNextPageToken()) {
+                // Adds the new page token found.
+                $pageTokens[] = $response->getPage()->getNextPageToken();
+            } else {
+                // Changes the requested page for the latest page that exists if the new page token
+                // found is empty (the requested page does not exist).
+                $pageNo = count($pageTokens);
+            }
+        }
+
+        // Fetches the actual page to display results of.
+        $response = $googleAdsClient->getGoogleAdsServiceClient()->search(
+            $customerId,
+            $query,
+            [
+                'pageSize' => $entriesPerPage,
+                // Requests to return the total results count. This is necessary to
+                // determine how many pages of results exist.
+                'returnTotalResultsCount' => true,
+                'pageToken' => $pageTokens[$pageNo - 1]
+            ]
+        );
+        if ($response->getPage()->getNextPageToken()) {
+            // Adds the new page token if there is one (it is not the last page).
+            $pageTokens[] = $response->getPage()->getNextPageToken();
+        }
```

Comment:
Same question: could you double-check if this will advance the pointer twice?

Reply:
The code example works as intended, there is no "pointer" used here per se. What we do is to fetch pages based on which one has to be displayed and store page tokens along the way for potential future use.